### PR TITLE
Bspline: 3d_to_2d with dimension

### PR DESCRIPTION
**Pull request by sbendjebla**
* **Please check if the PR fulfills these requirements**
- [ ] The commit messages follows our guidelines (explicit, in english)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] CHANGELOG.md has been updated

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
- [x] Bug fix:
- [ ] New features:
- [ ] Documentation
- [ ] Optimization

* **Other information**:
 _I improved the methods for **3d_to_2d_with_dimension**, and **2d_with_dimension_to_3d** (for edges: linesegment, bsplinecurve, arc/ wires/ contours)._
_I improved **arc3d_to_2d**: the result is now a **bsplinecurve2d** instead of a list of linesegment2d_
_I added **arc2d_to_3d**: the result is a **bsplinecurve3d**_

### CODE REVIEW

File: volmdlr/faces.py [hunk 1]
```diff
@@ -2710,19 +2720,200 @@ def point2d_with_dimension_to_3d(self, point2d, points_x, points_y, xmin, xmax,
...
+        return linesegment2d_with_dimension
+
+
+    def linesegment2d_with_dimension_to_parametric_frame(self, linesegment2d):            
+        '''
+        convert a linesegment2d from the dimensioned to the parametric frame
+        '''
+
+        for key in self._grids2d:
+            [points_x, points_y, xmin, xmax, ymin, ymax] = key
```
Comment (AxelRinghausen) [inline volmdlr/faces.py:2752]: You are overwriting points_x, points_y, xmin, xmax, ymin, ymax each time you iterate on the loop. So it turns out to be equal to `[points_x, points_y, xmin, xmax, ymin, ymax] = self._grids2d[-1]`

File: volmdlr/faces.py [hunk 2]
```diff
@@ -2710,19 +2720,200 @@ def point2d_with_dimension_to_3d(self, point2d, points_x, points_y, xmin, xmax,
...
+
+    def bsplinecurve2d_with_dimension_to_parametric_frame(self, bsplinecurve2d):
+        '''
+        convert a bsplinecurve2d from the dimensioned to the parametric frame
+        '''
+
+        points_dim = bsplinecurve2d.control_points
+        points = []
+        for key in self._grids2d:
+            [points_x, points_y, xmin, xmax, ymin, ymax] = key
```
Comment (AxelRinghausen) [inline volmdlr/faces.py:2816]: Same comment.

File: volmdlr/faces.py [hunk 3]
```diff
@@ -2710,19 +2720,200 @@ def point2d_with_dimension_to_3d(self, point2d, points_x, points_y, xmin, xmax,
...
+
+        xmin, xmax, ymin, ymax = 0, 1, 0, 1
+
+        start = self.point2d_parametric_to_dimension(arc2d.start, points_x, points_y, xmin, xmax, ymin, ymax)
+        end = self.point2d_parametric_to_dimension(arc2d.end, points_x, points_y, xmin, xmax, ymin, ymax)
+        interior = self.point2d_parametric_to_dimension(arc2d.interior, points_x, points_y, xmin, xmax, ymin, ymax)
+
+        arc2d_with_dimension = volmdlr.edges.Arc2D(start, interior, end)
+
+        return arc2d_with_dimension
```
Comment (AxelRinghausen) [inline volmdlr/faces.py:2853]: Are you sure an Arc2D in parametric coordinates is also an Arc2D in dimensional coordinates ?

File: volmdlr/faces.py [hunk 4]
```diff
@@ -2710,19 +2720,200 @@ def point2d_with_dimension_to_3d(self, point2d, points_x, points_y, xmin, xmax,
...
+        for key in self._grids2d:
+            [points_x, points_y, xmin, xmax, ymin, ymax] = key
+
+        start = self.point2d_with_dimension_to_parametric_frame(arc2d.start, points_x, points_y, xmin, xmax, ymin, ymax)
+        interior = self.point2d_with_dimension_to_parametric_frame(arc2d.interior, points_x, points_y, xmin, xmax, ymin, ymax)
+        end = self.point2d_with_dimension_to_parametric_frame(arc2d.end, points_x, points_y, xmin, xmax, ymin, ymax)
+
+        arc2d = volmdlr.edges.Arc2D(start, interior, end)
+
+        return arc2d
```
Comment (AxelRinghausen) [inline volmdlr/faces.py:2881]: The other way around now : are you sure an Arc2D in dimensional coordinates is also an Arc2D in parametric coordinates ?

File: volmdlr/faces.py [hunk 5]
```diff
@@ -3098,37 +3298,44 @@ def edge3d_to_2d_with_dimension(self, edge3d, points_x, points_y):
...
-        
-        edges2d = []
-        for edge in wire3d.primitives:
-            edges2d.append(volmdlr.edges.LineSegment2D(self.point3d_to_2d_with_dimension(edge.start, points_x, points_y, xmin, xmax, ymin, ymax),
-                                                      self.point3d_to_2d_with_dimension(edge.end, points_x, points_y, xmin, xmax, ymin, ymax)))
-        
-        return volmdlr.wires.Wire2D(edges2d)
+        for key in self._grids2d:
+            points_x = key[0]
+            points_y = key[1]
```
Comment (AxelRinghausen) [inline volmdlr/faces.py:3334]: Same comment here about self._grids2d.

File: volmdlr/faces.py [hunk 6]
```diff
@@ -2182,12 +2182,22 @@ def bsplinecurve3d_to_2d(self, bspline_curve3d):
...
+        return [vme.BSplineCurve2D.from_points_interpolation(
+                    points, max(self.degree_u, self.degree_v))]
+
+    def arc2d_to_3d(self, arc2d):
+        number_points = math.ceil(arc2d.angle * 7) + 1  # 7 points per radian
+        l = arc2d.length()
+        points = [self.point2d_to_3d(arc2d.point_at_abscissa(
+            i * l / (number_points - 1))) for i in range(number_points)]
+        return vme.BSplineCurve3D.from_points_interpolation(
+                    points, max(self.degree_u, self.degree_v))
```
[the remark this answers is not in the document]
Comment (sbendjebla) [reply volmdlr/faces.py:2200]: It's done !

**Review by AxelRinghausen — CHANGES_REQUESTED** [review]

File: volmdlr/faces.py [hunk 7]
```diff
@@ -2710,19 +2732,199 @@ def point2d_with_dimension_to_3d(self, point2d, points_x, points_y, xmin, xmax,
...
+
+        [points_x, points_y, xmin, xmax, ymin, ymax] = self._grids2d[0]
+
+        start = self.point2d_with_dimension_to_parametric_frame(arc2d.start, points_x, points_y, xmin, xmax, ymin, ymax)
+        interior = self.point2d_with_dimension_to_parametric_frame(arc2d.interior, points_x, points_y, xmin, xmax, ymin, ymax)
+        end = self.point2d_with_dimension_to_parametric_frame(arc2d.end, points_x, points_y, xmin, xmax, ymin, ymax)
+
+        arc2d = volmdlr.edges.Arc2D(start, interior, end)
+
+        return arc2d
```
[the remark this answers is not in the document]
Comment (sbendjebla) [reply volmdlr/faces.py]: It's done !

**Review by AxelRinghausen — APPROVED** [review]

**Pull request merged**